c/tskit/tables.c

```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        ret = TSK_ERR_SITE_OUT_OF_BOUNDS;
+        goto out;
+    }
+    metadata_length = source->metadata_offset[row + 1] - source->metadata_offset[row];
+    ancestral_state_length
+        = source->ancestral_state_offset[row + 1] - source->ancestral_state_offset[row];
+    ret = tsk_site_table_add_row(dest, source->position[row],
+        source->ancestral_state + source->ancestral_state_offset[row],
+        ancestral_state_length, source->metadata + source->metadata_offset[row],
+        metadata_length);
+    if (ret < 0) {
+        goto out;
+    }
+out:
+    return ret;
+}
+
+static int TSK_WARN_UNUSED
+migration_table_copy_row(tsk_migration_table_t *source, tsk_migration_table_t *dest,
+    tsk_id_t row, tsk_id_t *node_map, tsk_id_t *population_map)
+{
+    int ret = 0;
+    tsk_size_t metadata_length;
+    if (row >= (tsk_id_t) source->num_rows) {
+        ret = TSK_ERR_MIGRATION_OUT_OF_BOUNDS;
+        goto out;
+    }
+    metadata_length = source->metadata_offset[row + 1] - source->metadata_offset[row];
+    ret = tsk_migration_table_add_row(dest, source->left[row], source->right[row],
+        node_map[source->node[row]], population_map[source->source[row]],
+        population_map[source->dest[row]], source->time[row],
+        source->metadata + source->metadata_offset[row], metadata_length);
+    if (ret < 0) {
+        goto out;
+    }
+out:
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_table_collection_subset(
+    tsk_table_collection_t *self, tsk_id_t *nodes, size_t num_nodes)
+{
+    int ret = 0;
+    tsk_id_t k, n, m, ind, pop, mut, site;
+    tsk_id_t *node_map, *individual_map, *population_map, *site_map, *mutation_map;
+    tsk_table_collection_t tables;
+
+    node_map = calloc(self->nodes.num_rows, sizeof(tsk_id_t));
+    individual_map = calloc(self->individuals.num_rows, sizeof(tsk_id_t));
+    population_map = calloc(self->populations.num_rows, sizeof(tsk_id_t));
+    site_map = calloc(self->sites.num_rows, sizeof(tsk_id_t));
+    mutation_map = calloc(self->mutations.num_rows, sizeof(tsk_id_t));
+    if (node_map == NULL || individual_map == NULL || population_map == NULL
+        || site_map == NULL || mutation_map == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(node_map, 0xff, self->nodes.num_rows * sizeof(tsk_id_t));
+    memset(individual_map, 0xff, self->individuals.num_rows * sizeof(tsk_id_t));
+    memset(population_map, 0xff, self->populations.num_rows * sizeof(tsk_id_t));
+    memset(site_map, 0xff, self->sites.num_rows * sizeof(tsk_id_t));
+    memset(mutation_map, 0xff, self->mutations.num_rows * sizeof(tsk_id_t));
+
+    ret = tsk_table_collection_copy(self, &tables, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_clear(self);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        n = nodes[k];
```

Comment:
I find the notation ``n`` for a node confusing. We use variables like ``u`` and ``v`` to refer to nodes elsewhere, and keep ``n`` for things like "this is a fixed number".